app/uk/gov/hmrc/transitmovementsvalidator/services/XmlValidationService.scala

```diff
@@ -142,47 +157,129 @@ class XmlValidationServiceImpl @Inject() (implicit ec: ExecutionContext) extends
+                            withinCustomsOfficeElements = true
+                          }
+                        }
+
+                        if (withinCustomsOfficeElements && qName.equals("referenceNumber")) {
+                          inReferenceNumber = true
+                        }
                       }
 
-                      override def endElement(uri: String, localName: String, qName: String): Unit =
+                      override def endElement(uri: String, localName: String, qName: String): Unit = {
                         if (qName.equals("messageType")) {
                           inMessageTypeElement = false
                         }
 
+                        if (checkedMessageType.equals("OfficeOfDestinationActual")) {
+                          if (qName.equals("CustomsOfficeOfDestinationActual")) {
+                            inCustomsOfficeOfDestinationActual = false
+                            withinCustomsOfficeDestinationActual = false
+                            withinCustomsOfficeElements = false
+                          }
+                        }
+
+                        if (checkedMessageType.equals("OfficeOfDeparture")) {
+
+                          if (qName.equals("CustomsOfficeOfEnquiryAtDeparture") || qName.equals("CustomsOfficeOfDeparture")) {
+                            inCustomsOfficeOfDeparture = false
+                            withinCustomsOfficeElements = false
+                            withinCustomsOfficeEnquiryAtDeparture = false
+                          }
+
+                        }
+
+                        if (withinCustomsOfficeElements && qName.equals("referenceNumber")) {
+                          inReferenceNumber = false
+
+                          if (withinCustomsOfficeDestinationActual) {
+                            customsOfficeOfDestinationActualReferenceNumber = Some(referenceNumber)
+                          }
+
+                          if (withinCustomsOfficeEnquiryAtDeparture) {
+                            customsOfficeOfEnquiryAtDepartureReferenceNumber = Some(referenceNumber)
+                          }
+                        }
+
+                      }
+
                       override def startPrefixMapping(prefix: String, uri: String): Unit =
                         startPrefix = prefix
 
                     }
                   )
 
-                  if (!elementValue.equalsIgnoreCase(rootTag)) {
+                  val rootNodeCheck = if (!elementValue.equalsIgnoreCase(rootTag.split(":")(1))) {
                     Either.left(BusinessValidationError("Root node doesn't match with the messageType"))
                   } else {
                     Either.right()
                   }
 
+                  val referenceNumberCheck = if (!referenceNumber.toUpperCase.startsWith("GB") && !referenceNumber.toUpperCase.startsWith("XI")) {
+                    Either.left(BusinessValidationError(s"Invalid reference number: $referenceNumber"))
```

Comment:
We already do this check in the router (the ticket is to bring it to here), so the error message should match that in the router. See:

https://github.com/hmrc/transit-movements-router/blob/93ca96465f3d8d6de2ed6b453744dffecd0b4892/app/uk/gov/hmrc/transitmovementsrouter/services/CustomOfficeExtractorService.scala#L73

Specifically, we also need to ensure we tell the trader which element needs changing.

Reply:
Done